File: app/android/app/src/main/java/org/broadinstitute/ebolarisk/DataHolder.java
package org.broadinstitute.ebolarisk;

import android.content.res.AssetManager;
import android.util.Log;

import java.util.HashMap;

// What's the best way to share data between activities?
// http://stackoverflow.com/a/4878259
public class DataHolder {
  private LogRegModel minModel;
  private LogRegModel fullModel;
  private LogRegModel wellModel;
  private LogRegModel selModel;
  private HashMap<String, Float> floatVars;
  private HashMap<String, Float> floatExtra;
  private static DataHolder instance = new DataHolder();

  public static DataHolder getInstance() {
    return instance;
  }
  private DataHolder() {
    minModel = null;
    fullModel = null;
    wellModel = null;
    selModel = null;
    floatVars = new HashMap<String, Float>();
    floatExtra = new HashMap<String, Float>();
  }

  public static void loadModels(AssetManager am) {
    try {
      instance.minModel = new LogRegModel(am.open("min/mice.txt"), am.open("min/minmax.txt"));
      System.out.println("PRESENTATION-MIN MODEL ========================");
      System.out.println(instance.minModel);
    } catch (java.io.IOException ex) {
      Log.e("EbolaCARE2", "Cannot load the min model");
    }

    try {
      instance.fullModel = new LogRegModel(am.open("kgh/mice.txt"), am.open("kgh/minmax.txt"));
      System.out.println("PRESENTATION-FULL MODEL ========================");
      System.out.println(instance.fullModel);
    } catch (java.io.IOException ex) {
      Log.e("EbolaCARE2", "Cannot load the full model");
    }

    try {
      instance.wellModel = new LogRegModel(am.open("well/mice.txt"), am.open("well/minmax.txt"));
      System.out.println("WELLNESS MODEL ========================");
      System.out.println(instance.wellModel);
    } catch (java.io.IOException ex) {
      Log.e("EbolaCARE2", "Cannot load the wellness model");
    }

    // no model currently selected.
    instance.selModel = null;
  }

  public void setData(String var, boolean value) {
    Log.e("----->", var + ": " + value);
    if (value) floatVars.put(var, 1.0f);
    else floatVars.put(var, 0.0f);
    selectModel();
  }

  public void setData(String var, float value) {
    Log.e("----->", var + ": " + value);
    if (Float.isNaN(value)) {
      floatVars.remove(var);
    } else {
      floatVars.put(var, value);
    }
    selectModel();
  }

  public void remData(String var) {
    Log.e("----->", "Removed variable " + var);
    floatVars.remove(var);
    selectModel();
  }

  public void setDataExtra(String var, float value) {
    Log.e("----->", var + ": " + value);
    if (Float.isNaN(value)) {
      floatExtra.remove(var);
    } else {
      floatExtra.put(var, value);
    }
  }

  private void selectModel() {
    if (fullModel.contains(floatVars.keySet())) {
      System.out.println("===================== Selected Full Presentation Model");
      selModel = fullModel;
    } else if (wellModel.contains(floatVars.keySet())) {
      System.out.println("===================== Selected Wellness Presentation Model");
      selModel = wellModel;
    } else if (minModel.contains(floatVars.keySet())) {
      System.out.println("===================== Selected Minimal Presentation Model");
      selModel = minModel;
    } else {
      selModel = null;
    }
  }

  public LogRegModel getSelectedModel() {
    return selModel;
  }

  public HashMap<String, Float> getCurrentData() {
    return floatVars;
  }

  public HashMap<String, Float> getExtraData() {
    return floatExtra;
  }
}

File: app/ios/ebolarisk/DataHolder.swift
//
//  DataHolder.swift
//  ebolarisk
//
//  Created by Andres Colubri on 4/15/18.
//  Copyright © 2018 broadinstitute. All rights reserved.
//

import Foundation

// Using singletons for global data access
// https://medium.com/if-let-swift-programming/the-swift-singleton-pattern-442124479b19
// https://stackoverflow.com/questions/38793601/singleton-usage-in-swift
final class DataHolder {
    var minModel: LogRegModel
    var fullModel: LogRegModel
    var wellModel: LogRegModel
    var floatVars: [String: Float]
    var floatExtra: [String: Float]
    
    static let instance = DataHolder()
    
    var hasModel = false
    var selModel: LogRegModel
    
    private init() {
        minModel = LogRegModel(modelName: "min")
        fullModel = LogRegModel(modelName: "kgh")
        wellModel = LogRegModel(modelName: "well")
        
        print("MINIMAL MODEL ========================")
        print(minModel.toString())
        print("FULL MODEL ========================")
        print(fullModel.toString())
        print("WELLNESS MODEL ========================")
        print(wellModel.toString())
        
        selModel = minModel
        
        floatVars = [:]
        floatExtra = [:]
    }
    
    func setData(varName: String, value: Bool) {
        print("-----> " + varName + ": " + String(value))
        if value {
            floatVars[varName] =  1.0
        } else {
            floatVars[varName] =  0.0
        }
        selectModel()
    }

    func setData(varName: String, value: Float) {
        print("-----> " + varName + ": " + String(value))
        if value.isNaN {
            floatVars.removeValue(forKey: varName)
        } else {
            floatVars[varName] = value
        }
        selectModel()
    }
    
    func remData(varName: String) {
        print("-----> Removed variable " + varName)
        floatVars.removeValue(forKey: varName)
        selectModel()
    }
    
    func setDataExtra(varName: String, value: Float) {
        print("-----> " + varName + ": " + String(value))
        if value.isNaN {
            floatExtra.removeValue(forKey: varName)
        } else {
            floatExtra[varName] = value
        }
    }
    
    func selectModel() {
        if fullModel.contains(vars: Array(floatVars.keys)) {
            print("===================== Selected Full Presentation Model")
            selModel = fullModel
            hasModel = true
        } else if wellModel.contains(vars: Array(floatVars.keys)) {
            print("===================== Selected Wellness Presentation Model")
            selModel = wellModel
            hasModel = true
        } else if minModel.contains(vars: Array(floatVars.keys)) {
            print("===================== Selected Minimal Presentation Model")
            selModel = minModel
            hasModel = true
        } else {
            print("===================== Not enough data to select a model")
            hasModel = false
        }
    }
    
    func getSelectedModel() -> LogRegModel? {
        if hasModel {
            return selModel
        } else {
            return nil
        }
    }
    
    func getCurrentData() -> [String: Float] {
        return floatVars
    }
    
    func getExtraData() -> [String: Float] {
        return floatExtra
    }
}
